src/browser/modules/Sidebar/ProjectFiles.tsx

```diff
@@ -0,0 +1,171 @@
+          setError('')
+        }
+      })
+    // when a non-Project File action sets content in the editor
+    bus &&
+      bus.take(SET_CONTENT, () => {
+        if (isStillMounted) {
+          setSaveMode(false)
+          setFileName('')
+          setFileContents('')
+          setError('')
+        }
+      })
+    return () => {
+      isStillMounted = false
+    }
+  })
+
+  useEffect(() => {
+    // send mounted/unmounted message
+    bus && bus.send(PROJECT_FILES_MOUNTED, '')
+    return () => bus && bus.send(PROJECT_FILES_UNMOUNTED, '')
+  }, [])
+
+  return (
+    <Drawer id="ProjectFiles">
+      <DrawerHeader>Project Files</DrawerHeader>
+      <Render if={isSaveMode}>
+        <StyledSetting>
+          <StyledSettingLabel title={'Enter a file name'}>
+            {'File Name'}
+            <StyledSettingTextInput
+              onChange={event => {
+                setFileName(event.target.value)
+              }}
+              value={fileName}
+            />
+          </StyledSettingLabel>
+        </StyledSetting>
+        <a
+          onClick={async () => {
+            if (!fileName.length) {
+              setError('File name cannot be empty')
+              return
+            }
+            if (fileName.includes('/') || fileName.includes('\\')) {
+              setError('File name cannot include / or \\')
+              return
+            }
+            if (fileName.length) {
+              console.log('filename', fileName)
+              try {
+                const {
+                  data: { addProjectFile }
+                } = await addFile({
+                  variables: {
+                    projectId: 'project-03688c10-a811-4c0c-85d4-581e88c2183a', // @todo: hardcoded for now
+                    fileUpload: new File(
+                      [fileContents],
+                      `${fileName}${CYPHER_FILE_EXTENSION}`
+                    )
+                  },
+                  update: updateCacheAddProjectFile
+                })
+                const addedFile = {
+                  name: addProjectFile.name,
+                  directory: addProjectFile.directory,
+                  extension: addProjectFile.extension
+                }
+                setSaveMode(false)
+                setFileName('')
+                setFileContents('')
+                setError('')
+                bus.send(SELECT_PROJECT_FILE, addedFile) // set ProjectFileButton to edit mode
+              } catch (e) {
+                console.log('++here', e)
+              }
+            }
+          }}
+        >
+          Save
+        </a>
+        <a
+          onClick={() => {
+            setSaveMode(false)
+            setError('')
+          }}
+        >
+          Cancel
+        </a>
+      </Render>
+      <ProjectFilesError apolloErrors={[apolloError]} errors={[error]} />
+      <ProjectFilesScripts />
+    </Drawer>
+  )
+}
+
+const mapStateToProps = (state: any) => {
+  return state
```

Comment:
rather return nothing in this case, this will make the component re-render on any state change

Reply:
yeah I am going to do something with this in next PR so I left it, will remove otherwise.